contrib/mesos/pkg/node/registrator.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"reflect"
+	"time"
+
+	log "github.com/golang/glog"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	defaultInboxSize = 1000
+)
+
+type Registrator interface {
+	Register(hostName string, labels map[string]string) (bool, error)
+	Run(terminate <-chan struct{}) error
+}
+
+type registration struct {
+	hostName string
+	labels   map[string]string
+}
+
+func (r *registration) Copy() queue.Copyable {
+	return &registration{
+		hostName: r.hostName,
+		labels:   r.labels, // labels are never changed, hence just copy the slice
+	}
+}
+
+func (r *registration) GetUID() string {
+	return r.hostName
+}
+
+func (r *registration) Value() queue.UniqueCopyable {
+	return r
+}
+
+func (r *registration) Is(types queue.EventType) bool {
+	return types&queue.UPDATE_EVENT != 0
+}
+
+type clientRegistrator struct {
+	store  cache.Store
+	client *client.Client
+	queue  *queue.HistoricalFIFO
+	inbox  chan queue.Entry
+}
+
+func NewRegistrator(client *client.Client, store cache.Store) *clientRegistrator {
+	inbox := make(chan queue.Entry, defaultInboxSize)
+	return &clientRegistrator{
+		store:  store,
+		client: client,
+		inbox:  inbox,
+		queue:  queue.NewHistorical(inbox),
+	}
+}
+
+func (r *clientRegistrator) Run(terminate <-chan struct{}) error {
+	loop := func() {
+	RegistrationLoop:
+		for {
+			obj := r.queue.CancelablePop(terminate)
+			if obj == nil {
+				break RegistrationLoop
+			}
+			select {
+			case <-terminate:
+				break RegistrationLoop
+			default:
+			}
+
+			rg := obj.(*registration)
+			n, upToDate, err := r.isUpToDate(rg.hostName, rg.labels)
+			if err != nil {
+				log.Error(err)
```

Comment:
why do we continue in this case? isn't err != nil the equiv to found == false? and in that case, why even return an error from isUpToDate because it means the same thing to the caller as not-found, right?